retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/StreamingCallSubscriber.java

```diff
@@ -0,0 +1,186 @@
+                responseBuilder.message(status.reasonPhrase());
+            }
+
+            headers.forEach(header -> responseBuilder.addHeader(header.getKey().toString(),
+                                                                header.getValue()));
+
+            if (contentType == null) {
+                contentType = headers.get(HttpHeaderNames.CONTENT_TYPE);
+            }
+
+            if (contentLength == NO_CONTENT_LENGTH) {
+                contentLength = headers.getLong(HttpHeaderNames.CONTENT_LENGTH, NO_CONTENT_LENGTH);
+            }
+            return;
+        }
+        if (!callbackCalled) {
+            responseBuilder.body(ResponseBody.create(
+                    Strings.isNullOrEmpty(contentType) ? null : MediaType.parse(contentType),
+                    contentLength, Okio.buffer(new ForwardingSource(pipeBuffer.source()) {
+                        @Override
+                        public long read(Buffer sink, long byteCount) throws IOException {
+                            subscription.request(1);
+                            return super.read(sink, byteCount);
+                        }
+                    })));
+            responseBuilder.request(request);
+            responseBuilder.protocol(Protocol.HTTP_1_1);
+            safeOnResponse(responseBuilder.build());
+        }
+
+        final HttpData data = (HttpData) httpObject;
+        pipeBuffer.write(data.array(), data.offset(), data.length());
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        if (armeriaCall.tryFinish()) {
+            safeOnFailure(new IOException(throwable.getMessage(), throwable));
+        } else {
+            safeOnFailure(newCanceledException());
+        }
+        pipeBuffer.close(throwable);
+    }
+
+    @Override
+    public void onComplete() {
+        if (!armeriaCall.tryFinish()) {
+            safeOnFailure(newCanceledException());
+        }
+        pipeBuffer.close(null);
+    }
+
+    private void onCancel(Subscription subscription) {
+        final IOException canceledException = newCanceledException();
+        safeOnFailure(canceledException);
+        subscription.cancel();
+        pipeBuffer.close(canceledException);
+    }
+
+    private void safeOnFailure(IOException e) {
+        if (callbackCalled) {
+            return;
+        }
+        callbackCalled = true;
+        callbackExecutor.execute(() -> callback.onFailure(armeriaCall, e));
+    }
+
+    private void safeOnResponse(Response response) {
+        if (callbackCalled) {
+            return;
+        }
+        callbackCalled = true;
+        callbackExecutor.execute(() -> {
+            try {
+                callback.onResponse(armeriaCall, response);
+            } catch (IOException e) {
+                callback.onFailure(armeriaCall, e);
+            }
+        });
+    }
+
+    private static IOException newCanceledException() {
+        return new IOException("Canceled");
```

Comment:
- `cancelled`? (To be consistent with `CompletableFuture.isCancelled()`)
- `newCancelledException`